### ISSUE TITLE
FormBuilder cannot be usefully inherited

### ISSUE THREAD

**Issue by danni** [wagtail/wagtail#3198]
`wagtail.wagtailforms.forms.FormBuilder` uses unbound references to refer to its methods, which means you can't inherit the form builder to override a method. Suggest using the bound methods.

It would also be good to abstract away common code like building the choices array, to allow this to be consistently overridden.

**Comment by lb-** [comment]
@danni I know it's been a while since you posted this but it would be good to get some more context on what you are trying to achieve and what blockages there are.

I have added fields types to forms reasonably easily, it could be smoother but I did not run into any major issues.

There is a rough edge in extending the FORM_FIELD_CHOICES, maybe there could be a method such as `get_form_field_choices` on `AbstractFormField`.

**Example**
Note: Does not include any of the parsing of the uploaded image file and storing it but could work for any other field type.

```
EXTENDED_FORM_FIELD_CHOICES = (
    ('singleline', _('Single line text')),
    ('multiline', _('Multi-line text')),
    ('email', _('Email')),
    ('number', _('Number')),
    ('url', _('URL')),
    ('checkbox', _('Checkbox')),
    ('checkboxes', _('Checkboxes')),
    ('dropdown', _('Drop down')),
    ('multiselect', _('Multiple select')),
    ('radio', _('Radio buttons')),
    ('date', _('Date')),
    ('datetime', _('Date/time')),
    ('image', _('Upload Image')),
)

class FormField(AbstractFormField):
    """Stores a single field with custom information such as label."""

    page = ParentalKey('FormPage', related_name='form_fields')

    # Override the field type choices
    field_type = models.CharField(
        verbose_name=_('field type'),
        max_length=16,
        choices=EXTENDED_FORM_FIELD_CHOICES)


class ExtendedFormBuilder(FormBuilder):
    """
        Create a form instance from Fields defined in :model:`main.FormPage`.
        Adds a special new field to enable image uploading.
    """

    def create_image_upload_field(self, field, options):
        """Retun an instance of a WagtailImageField."""
        return WagtailImageField(**options)

    # change the field types available, each key maps to a function
    FIELD_TYPES = FormBuilder.FIELD_TYPES
    FIELD_TYPES.update({'image': create_image_upload_field})


class FormPage(AbstractEmailForm):
    """Stores a page where users can define their own form."""
    form_builder = ExtendedFormBuilder
    # ... rest of the fiels such as intro & additional methods

```

**Comment by danni** [comment]
The problem I wanted to solve was replacing an existing `create_` method, which I wanted to chain up to the parent method, and thus required quite a bit of fiddling to get my references in alignment.

Additionally, I wanted to reuse the choices array in a number of similar widgets, which required reimplementing it.

**Comment by lb-** [comment]
@danni makes sense, I have tested that out and understand now, would love your thoughts on a potential solution below.

**In [wagtailforms/forms.py](https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailforms/forms.py)**
1. Remove `FIELD_TYPES` all together.
2. Rely on correctly formatted method names like `create_singleline_field` to determine if the field type is valid. (this might be a bit risky)
3. Add a new `get_create_field_function` method, also adding to the exception with more helpful info for developers.
```python
    def get_create_field_function(self, type):
        """
            Takes string of field type returns a Django Form Field Instance.
            Assumes form field creation functions are in the format:
            'create_fieltype_field'
        """
        create_field_function = getattr(self, 'create_%s_field' % type, None)
        if create_field_function:
            return create_field_function
        else:
            import inspect
            method_list = [
                f[0] for f in
                inspect.getmembers(self.__class__, inspect.isfunction)
                if f[0].startswith('create_') and f[0].endswith('_field')
            ]
            raise Exception(
                "Could not find function matching format \
                create_<fieldname>_field for type: " + type,
                "Must be one of: " + ", ".join(method_list)
            )

```
3. Replace the [`formfields` property](https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailforms/forms.py#L98-L110) with this code:
```python
    @property
    def formfields(self):
        formfields = OrderedDict()

        for field in self.fields:
            options = self.get_field_options(field)
            create_field = self.get_create_field_function(field.field_type)
            formfields[field.clean_name] = create_field(field, options)

        return formfields
```

**In [wagtailforms/models.py](https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailforms/models.py)**
4. Add a new method to `AbstractFormField` called `get_form_field_choices` that simply returns the current `FORM_FIELD_CHOICES`, this makes it easy to append or rewrite as needed without having to redefine the `field_type` model field itself.

**Comment by danni** [comment]
This looks very similar to what I ended up with when reimplementing all the things.

**Comment by lb-** [comment]
Implemented most of what I put in the notes, but did not add `get_form_field_choices` to `AbstractFormField` as that has other odd implications.

Docs in PR cover the way to add fields and it is pretty straight forward now.

**Comment by lb-** [comment]
Implemented in #4076
